File: src/main/python/rag/embedder.py
"""
Bedrock Titan Embeddings v2 wrapper.

embed_text — convert a text string to a dense float vector via Bedrock.

Accepts an optional bedrock_client kwarg so tests can inject a mock
without AWS credentials — identical pattern to sqs_to_s3.read_s3_json.
"""

import json
import logging

from core.aws import bedrock_runtime_api
from core.config import config as _config_fn

log = logging.getLogger(__name__)

_cfg = _config_fn()
_EMBED_MODEL: str = _cfg.get("rag", "embed_model", fallback="amazon.titan-embed-text-v2:0")
_EMBED_DIMS: int = int(_cfg.get("rag", "embed_dims", fallback="1024"))


def embed_text(text: str, *, bedrock_client=None) -> list[float]:
    """Embed a single text string using Titan Embeddings v2.

    Args:
        text: The text to embed. Titan v2 supports up to 8192 tokens.
        bedrock_client: Optional pre-built boto3 bedrock-runtime client.

    Returns:
        List of floats of length _EMBED_DIMS (default 1024).
    """
    client = bedrock_client or bedrock_runtime_api()
    body = json.dumps({"inputText": text, "dimensions": _EMBED_DIMS})
    response = client.invoke_model(
        modelId=_EMBED_MODEL,
        contentType="application/json",
        accept="application/json",
        body=body,
    )
    result = json.loads(response["body"].read())
    log.debug("Embedded %d chars → %d-dim vector", len(text), len(result["embedding"]))
    return result["embedding"]

File: src/main/java/com/cloudpid/ai/rag/Embedder.java
package com.cloudpid.ai.rag;

import com.cloudpid.ai.config.AppConfig;
import com.cloudpid.ai.config.AppConfigFactory;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.bedrockruntime.BedrockRuntimeClient;
import software.amazon.awssdk.services.bedrockruntime.model.InvokeModelRequest;
import software.amazon.awssdk.services.bedrockruntime.model.InvokeModelResponse;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Bedrock Titan Embeddings v2 wrapper.
 * Mirrors Python {@code rag/embedder.py}.
 */
public final class Embedder {

    private static final Logger log = LoggerFactory.getLogger(Embedder.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final BedrockRuntimeClient client;
    private final String modelId;
    private final int dims;

    public Embedder(BedrockRuntimeClient client, String modelId, int dims) {
        this.client = client;
        this.modelId = modelId;
        this.dims = dims;
    }

    public static Embedder create() {
        AppConfig cfg = AppConfigFactory.get();
        BedrockRuntimeClient client = BedrockRuntimeClient.builder()
            .region(Region.of(cfg.awsRegion()))
            .build();
        return new Embedder(client, cfg.ragEmbedModel(), cfg.ragEmbedDims());
    }

    /**
     * Embed a single text string using Titan Embeddings v2.
     * Mirrors Python {@code embed_text(text)}.
     */
    public List<Float> embedText(String text) {
        try {
            String body = MAPPER.writeValueAsString(Map.of("inputText", text, "dimensions", dims));
            InvokeModelResponse response = client.invokeModel(InvokeModelRequest.builder()
                .modelId(modelId)
                .contentType("application/json")
                .accept("application/json")
                .body(SdkBytes.fromUtf8String(body))
                .build());
            JsonNode result = MAPPER.readTree(response.body().asUtf8String());
            JsonNode embedding = result.path("embedding");
            if (embedding.isMissingNode() || !embedding.isArray()) {
                throw new IllegalStateException("Unexpected Bedrock embedding response: " + result);
            }
            List<Float> vector = new ArrayList<>();
            for (JsonNode val : embedding) {
                vector.add(val.floatValue());
            }
            log.debug("Embedded {} chars → {}-dim vector", text.length(), vector.size());
            return vector;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
